### PR TITLE
Added support for leading and trailing underscores and printing Dummy and Wild in LaTeX and pretty

### PR DESCRIPTION
**Pull request by oscargus**
#### References to other Issues or PRs

Fixes #6498 
Fixes #14718

#### Brief description of what is fixed or changed
Now, Dummy and WIld is printed with a leading and trailing underscore, respectively, in LaTeX and pretty printers.

Leading and trailing underscores in Symbol names are escaped in LaTeX and removed before later processing and restored in pretty.

#### Other comments
I didn't get WildFunction to work at this stage.

There is something strange with `Lambda`s with expressions in them. Or maybe it is supposed to be like that and I did not get it. 

```
def test_pretty_lambda():
    # S.IdentityFunction is a special case
    expr = Lambda(y, y)
    assert pretty(expr) == "_x -> _x"
    assert upretty(expr) == u"_x ↦ _x"

    expr = Lambda(x, x+1)
    assert pretty(expr) == "x -> x + 1"
    assert upretty(expr) == u"x ↦ x + 1"
```

For the first case it is printed like a Dummy, but not for the second case.

#### Release Notes

* printing
    * Leading and trailing underscores in symbol names are kept in LaTeX and pretty printer
    * `Dummy` and `Wild` are printed with leading and trailing underscore, respectively, in LaTeX and pretty printer, as in `str`

### CODE REVIEW

**Comment by oscargus** [comment]
Is there a good/simple way to encode the correct result string for the more complicated pretty printings? I.e. add line breaks and so on so it looks OK in the source.

**Comment by moorepants** [comment]
I like:

```python
var = """\
multi-line
line 2
line 3\
"""
```

**Comment by oscargus** [comment]
I wasn't that clear. If one looks at the summation pretty-print that I have changed so far,  I had to manually break after the \n and add a \ as the print-out was a single line with \n in it. That is sort of semi-feasible, if it wasn't for the editor adding spaces/tabbing based on spaces in the previous lines, the need to escape \, it is almost impossible to figure out where a space may be missing, and the fact that some of the failing tests are about ten lines.

Basically what I am looking for is some some of method/function to efficiently print "code" for one of these pretty-print strings. Code that I can simply copy-paste to the test and that still is readable (as opposed to the single-line output that one gets or the print version where the \n now are line-breaks). So basically something that escapes what needs to be escaped, add a \ and a line break and ideally the """ before and after.

**Comment by oscargus** [comment]
@asmeurer can you please have a look at this at your convenience. Is this what you expected from #6498 ?

**Comment by oscarbenjamin** [comment]
This patch looks good. I'll merge soon if no one objects.

**Comment by oscargus** [comment]
Thanks! I think it would be good if @asmeurer could have a say if this is what he expected from #6498 

Maybe I should add it for MathML as well, although it doesn't have to be this PR.

File: doc/src/modules/solvers/solveset.rst [hunk 1]
```diff
@@ -346,10 +346,10 @@ How do we manipulate and return an infinite solution?
    >>> from sympy import ImageSet, Lambda, pi, S, Dummy, pprint
    >>> n = Dummy('n')
    >>> pprint(ImageSet(Lambda(n, 2*pi*n), S.Integers), use_unicode=True)
-   {2⋅n⋅π | n ∊ ℤ}
+   {2⋅_n⋅π | _n ∊ ℤ}
```
Comment (smichr) [inline doc/src/modules/solvers/solveset.rst]: I don't think we want to do this. If there is no ambiguity in the expression's variables, just use symbols to print the dummies. There is ambiguity when any two symbols share the same name. In such cases, the names can be given arbitrary numbers to disambiguate them. Clashes can exist in multiple ways: 
```python
>>> len(set([Dummy('x'), Dummy('x'), Symbol('x', real=True), Symbol('x', integer=True), Wild('x', exclude=[y]), Wild('x', exclude=[z])])) == 6
True
```
Bound symbols should never appear with an underscore since (I believe) their context will make them unambiguous.

Comment (oscargus) [reply doc/src/modules/solvers/solveset.rst]: I somehow agree. It looks a bit weird. It is rather straightforward to reverse this. (At least from a code perspective, the documentation is a bit more of a hassle...) I just based it on https://github.com/sympy/sympy/issues/6498#issuecomment-461196106 so I think @asmeurer should have a say here.

Implementing adding of the underscore based on other names in the expression is far from trivial with the current `Printer` setup, since `_print_Dummy` does not have any idea about the expression.

Comment (smichr) [reply doc/src/modules/solvers/solveset.rst]: The user should not have to be bothered with seeing underscores if they don't add any information. So perhaps part of the printing process is to replace all Dummy/Wild that have a clashing name with Symbol *before* doing the printing. Underscores will only remain if there was a clash. (And if we don't like how it looks then we (as coder or user) need to pick variables that don't clash.
```python
def DummyWild_as_Symbol(eq):
    from sympy.core.symbol import Symbol, Dummy, Wild
    from collections import defaultdict
    free = eq.free_symbols
    bound = eq.atoms(Symbol) - free
    # replace bound symbols that are Dummy or Wild with Symbols
    # since these are never ambiguous in meaning
    eq = eq.subs(dict([(b, Symbol(b.name)) for b in bound if isinstance(b, (Dummy, Wild))]))
    # replace Dummy or Wild with Symbol if there is no ambiguity in doing so
    dups = defaultdict(list)
    for i in [i for i in free]:
        dups[i.name].append(i)
    for i in dups:
        if len(dups[i]) == 1:
            u = dups[i][0]
            if isinstance(u, (Dummy, Wild)):
                eq = eq.xreplace({u: Symbol(i)})
    return eq
```
User's who want to print `x + x` that looks nice (perhaps for pedagogical reasons) can do so with an unevaluated expression or can instantiate symbols with an unused assumption:
```
>>> x, y = [Symbol('x', _=i) for i in range(2)]
>>> x + y
x + x
```

Comment (smichr) [reply doc/src/modules/solvers/solveset.rst]: > user should not have to be bothered with seeing underscores

Revisiting this now, it seems that if we are seeing underscores appear then we should fix the offending code that put them there. In case of the ImageSet, a dummy should only be used if it clashes with an existing symbol.

Comment (oscargus) [reply doc/src/modules/solvers/solveset.rst]: > user should not have to be bothered with seeing underscores

I guess that is the main question here. According to #6498 that is exactly what the users should be bothered with. I do not have any strong opinions as such. Just solved the issue. @asmeurer ?

**Comment by oscarbenjamin** [comment]
This looks like a useful addition but sine it isn't fixing a recent regression I don't see why this needs to hold up the 1.5 release so removing the milestone
